### PR TITLE
Allow use scrubbing with remote execution (build actions with scrubbed inputs locally)

### PR DESCRIPTION
**Pull request by bozaro**
Since cannot run remote cache key scrubbing action with remote execution, run this actions only locally. Actions without scrubbed input can still be executed remotely.

Same idea as:
 - https://github.com/bazelbuild/bazel/pull/16240

### CODE REVIEW

File: src/main/java/com/google/devtools/build/lib/remote/RemoteExecutionService.java [hunk 1]
```diff
@@ -341,7 +341,8 @@ public CachePolicy getWriteCachePolicy(Spawn spawn) {
   public boolean mayBeExecutedRemotely(Spawn spawn) {
     return remoteCache instanceof RemoteExecutionCache
         && remoteExecutor != null
-        && Spawns.mayBeExecutedRemotely(spawn);
+        && Spawns.mayBeExecutedRemotely(spawn)
+        && !hasScrubbedInput(spawn, scrubber);
```
Comment (tjgq) [inline src/main/java/com/google/devtools/build/lib/remote/RemoteExecutionService.java:345]: I think it makes sense to allow remote execution for non-scrubbed actions, but I have some concerns with this implementation:

1. In addition to inputs, we might also scrub command line arguments and set a scrubbing salt, so this check is incomplete.
2. Expanding/iterating the inputs is a fairly expensive operation, and with this change we're going to do it twice for every matching action (first time here, second time when computing the actual cache key). I'd prefer to not make the decision to run remotely conditional on the input expansion.

Can we get away with simply `scrubber.forSpawn(spawn) == null` here? `Scrubber#forSpawn` is defined to return non-null iff at least one of the configuration rules matches the action; this is, strictly speaking, not a guarantee that scrubbing will occur (applying the respective transform might result in zero changes) but that ought to be the case with a carefully written configuration.

Comment (bozaro) [reply src/main/java/com/google/devtools/build/lib/remote/RemoteExecutionService.java:345]: > Can we get away with simply scrubber.forSpawn(spawn) == null here?

No, looks like I have non null `spawnScrubber` for every target with files input with configuration like:
```
rules {
  transform {
    omitted_inputs: "^bazel-out/volatile-status\\.txt$"
  }
}
```

Looks like I can create rules filter by set label or mnemonic pattern. But collecting these rules will be a very difficult task:
 - most stamping actions are declared in third-party code;
 - in case of an error, I get a quiet cache flush.

Comment (tjgq) [reply src/main/java/com/google/devtools/build/lib/remote/RemoteExecutionService.java:345]: In the use case for stamping that I'm most familiar with (build practices at Google), there's a relatively small number of targets that perform stamping - typically only one per language/ruleset. Each of these targets creates a single language-specific library embedding the workspace status, which may then get linked into every binary (or other top-level deployable artifact) for that language. This way, only a few, easy to identify targets depend on the workspace status files.

It sounds like your use of stamping is much more widespread, and can affect an unbounded number of targets, as opposed to a relatively small number of relatively "well-known" ones. Could you provide some more information on how your setup uses stamping?

Also - are you looking for a solution specifically for `volatile-status.txt`, or does this have to generalize to any input file? For example, suppose that action A consumes `volatile-status.txt` and produces an arbitrary file `a.out` (incorporating information from `volatile-status.txt`). Then a later action B consumes `a.out`. Would you then also need to scrub `a.out`?

Comment (bozaro) [reply src/main/java/com/google/devtools/build/lib/remote/RemoteExecutionService.java:345]: > Then a later action B consumes a.out. Would you then also need to scrub a.out?

Nope. Only `volatile-status.txt`.

We are trying to use stamping only for the top-level artifacts:
 - for executable files;
 - for docker images.

There are really few such goals.

The purpose of stamping is to save information in this file about which revision the file was compiled from. To do this, the revision number is transmitted via `volatile-status.txt`.

The problem is that stamping is enabled not for concrete targets, but globally via the `--stamp` flag. In some cases, it can affect, for example, on `bison` binary for generate `yacc` files in one of the third-party libraries we use. If for this target `volatile-status.txt` becomes significant, then we will get a cache miss for this and all dependent targets on every build. In this case, we get parasitic pressure on the cache, which is quite difficult to detect.

Comment (bozaro) [reply src/main/java/com/google/devtools/build/lib/remote/RemoteExecutionService.java:345]: Maybe, before implementing scrubbing support with remote build, it makes sense to do the following:
- use `scrubber.forSpawn(spawn) == null` cheap check for `mayBeExecutedRemotely' by default;
- use an expensive check with iteration over files enabling by the experimental flag like `--experimental_scrubbing-remote-build-expensive-check`.

Comment (tjgq) [reply src/main/java/com/google/devtools/build/lib/remote/RemoteExecutionService.java:345]: Sorry - I haven't forgotten about this, I just haven't had the time to prioritize it.

If possible, I'd like to avoid adding yet another flag (avoiding an explosion of flags is the reason we decided to go with a configuration file for scrubbing). I'd also like to keep the "matcher" and "transform" separate (i.e., determining whether scrubbing is enabled should depend only on the "matcher", not on the "transform").

Since this feature is only required to support the volatile status file (and not arbitrary files in the input), I think the right way to go might be something along these lines:

1. Add a `has_volatile_status_input` field to the `Matcher`.
2. Create a fast path (i.e., one that does not require doing a full expansion of the inputs) to determine whether a spawn has the volatile status as an input. For example, we could check directly for the presence of the volatile status file in the `InputMetadataProvider` for the spawn's owner, which can be done in O(1) instead of O(#inputs).

Comment (bozaro) [reply src/main/java/com/google/devtools/build/lib/remote/RemoteExecutionService.java:345]: Sorry for the pause, migrating to bazel 7.0 took more effort than I expected and I still have a project build that doesn't work.

I have collected some statistics on the use of `volatile-status.txt` within the go_binary rules (GoLink mnemonic): 539 rules of 875 (61.6%). Manually maintaining a list of targets for which inputs need to be filtered looks unrealistic.

As I undertant `has_volatile_status_input` field for `Matcher` can be used only when `omitted_inputs` ommits only `volatile-status.txt` rule. This looks like a very fragile check.

Also I want use `omitted_inputs` not only for `volatile-status.txt` status file.
For example:
 - Create docker `labels.json`-file based on `volatile-status.txt` (`volatile-status.txt` used as regular output and rule executes on every `volatile-status.txt` change, empty `omitted_inputs`);
 - Create docker image with labels from `labels.json` (`labels.json` used as volatile file and filtered by `omitted_inputs`).

Comment (fmeum) [reply src/main/java/com/google/devtools/build/lib/remote/RemoteExecutionService.java:345]: @bozaro Just curious, do you find that `go_binary` unnecessarily depends on `volatile-status.txt` or do you really want this many rules to depend on it?

Comment (bozaro) [reply src/main/java/com/google/devtools/build/lib/remote/RemoteExecutionService.java:345]: In addition, I hope that this check is a temporary solution and later scrubbing will be supported on the build farm.

Comment (bozaro) [reply src/main/java/com/google/devtools/build/lib/remote/RemoteExecutionService.java:345]: > Just curious, do you find that go_binary unnecessarily depends on volatile-status.txt or do you really want this many rules to depend on it?

I don't know about parasitic dependencies on `volatile-status.txt` inside the `go_binary'.

In our case, the main part of the executable files uses stamping to store a revision from which a similar file can be build.

Utilities that are used only inside the build job don't use stamping (in our case, the rules with stamping are not free, since they are linked locally, not on the farm). Although there are suspiciously many of them now. I'll figure it out.

**Comment by bozaro** [comment]
This PR works fine on `release-7.0.0rc2` and `release-7.1.0`, but on `7.0.0` tag it's requre patch like https://github.com/joomcode/bazel/commit/663beb61e4e7c8f708d3e058a670cc977d6a1362 (produced by cherry-pick https://github.com/bazelbuild/bazel/commit/2b700b5108e9969511ab8defe6612b5a50b4df86 after bisect between 7.0.0 and release-7.1.0).

**Comment by bozaro** [comment]
I am sorry... Rebasing to 7.0.0 tag was a bad idea...

File: src/main/java/com/google/devtools/build/lib/remote/RemoteExecutionService.java (hunk 1 above)
Comment (bozaro) [reply src/main/java/com/google/devtools/build/lib/remote/RemoteExecutionService.java:345]: I tried to imagine how scrubbing and remote build should work in the final form.

I came to the following conclusions:
 1. Global configuration via `--experimental_remote_scrubbing_config` is not scalable:
    - to specify scrubbing, you need to know mnemonic and omitted inputs, which depend on the implementation of the rules;
    - errors in scrubbing config are very difficult to detect (the build works, but just with a lower cache hit);
    - there is no clear way to find all the rules that require scrubbing config (especially in third-party libraries).
 2. The omitted inputs configuration must occur somewhere at the Starlark level of the rules implementation code. `volatile-status.txt ` should become a regular input.
 3. In the case when omitted inputs is configured at the rule level, you may need to use a check like `scrubber.forSpawn(spawn) == null`.
 4. Someday scrubbing will be supported by the Remote Execution Protocol and the problem will be solved definitively (but I think it will not be soon).

Thus, in the context of changing this PR, it turns out:
1. When the omitted inputs configuration appears at the Starlark code level of the rules, the check should look like: `scrubber.forSpawn(spawn) == null`
2. While there is no such check:
   - the current version of the check looks good enough;
   - since this `shouldOmitInput`-check will be deleted and it will be a breaking change, it needs to be hidden under the experimental flag like `--experimental_scrubbing-remote-build-expensive-check` (without the flag use `scrubber.forSpawn(spawn) == null` check);
   - a good starting point for `--experimental_remote_scrubbing_config` would be the file like:
     ```
     rules {
       transform {
         omitted_inputs: "^bazel-out/volatile-status\\.txt$"
       }
     }
     ```

@tjgq, what do you think about this?

Comment (bozaro) [reply src/main/java/com/google/devtools/build/lib/remote/RemoteExecutionService.java:345]: @tjgq, please let's decide something and finish this PR

Comment (bozaro) [reply src/main/java/com/google/devtools/build/lib/remote/RemoteExecutionService.java:345]: # Bazel 6 stamping vs remote cache/remote build

The original problem: I want to solve is that there is currently no way to use stamping and assembly on the farm without a permanent cache miss.

In the simplest case, a rule of the form is sufficient:
```
rules {
  transform {
    omitted_inputs: "^bazel-out/volatile-status\\.txt$"
  }
}
```
After that, the shared cache starts working the same way as the local one. This solves the problem of remote assembly for the current rules from Bazel 6 without knowledge of the internal logic of the rules.

Specifying the rules in the configuration is a very non-trivial task:
- I only know one way to find out the exact list of rules that use volatile input: add debugging output to my PR (for example: https://github.com/joomcode/bazel/commit/d956d63a66e64cb6aea802df38827c92272e0519?diff=unified&w=1). But it is not available for most developers.
- Dependency on volatile input can be transitive: for example, it is declared in `go_library`, and used in `go_binary'.
- The error is unnoticeable. It is very difficult to detect.
- The same mnemonic (`GoLink') may use stamping (toplevel-executables from `go_binary`), or it may not use stamping ('go_binary` tools, `go_test` tests, binaries for mock generation).

A universal rule that just works and does not require tuning looks very tempting.

If you limit the problem to just a file `bazel-out/volatile-status.txt` then it solves the problems that were there before.

# Bazel 6 stamping is broken

Scrubbing in Bazel 7 allows you to solve a very serious limitation of Bazel 6 stamping: volatile input must be used only in top level rules.

For example, I want to embed version information in the executable: this means that the data from the volatile input must be processed at the linking stage. If I, for example, convert `bazel-out/volatile-status.txt` to JSON, then I will get the same data in all executable files, and not the version that is needed to build the executable file.

Scrubbing allows you to get around this problem:
  - the rule for conversion should use `bazel-out/volatile-status.txt` as regular input;
  - linking should use the output of the conversion rule to JSON as a volatile input.

It's a shame to lose this feature with a remote build. It's like taking candy from a developer.

# Bazel 7 scrubbing configuration limitations

The full power of scrubbing is revealed if the decision about whether it is needed or not is made at the action level (above is an example with `GoLink`). How to solve this is a rather complicated question. I don't have an answer at the moment.

But now it is too easy to make a mistake in writing the rules and not notice it.


P.S. I'm sorry for my terrible English. I'm trying to convey my understanding as best I can.

Comment (tjgq) [reply src/main/java/com/google/devtools/build/lib/remote/RemoteExecutionService.java:345]: Alright, let's go through the problems one by one:

**Bazel 6 stamping vs remote cache/remote build**

I think this problem is easily solvable. If we implement the solution I outlined above, it becomes possible to write a single rule of the form:

```
rule {
  matcher {
    has_volatile_input: true
  }
  transform {
    omitted_inputs: "^bazel-out/volatile-status\\.txt$"
  }
}
```

and it will apply to every action that has `volatile-status.txt` in its inputs, without having to list those actions explicitly in the configuration. (There's a little bit of redundancy between `has_volatile_input` and `omitted_inputs`, but it's intentional: separating the matcher and the transform brings clarity and is a property I'd like to keep.)

However, the configuration above will **not** apply to an action that produces an output derived from `volatile-status.txt` (that's what the next problem is about).

** Bazel 6 stamping is broken **

It's unclear to me whether this is a bug in how stamping works, or an intended limitation. I don't mean to be dismissive; all I'm saying is that I didn't design the stamping feature and haven't used it all that much, so I can't authoritatively declare that there is a bug, and much less that cache key scrubbing is the correct way to solve it.

My suggestion here is to open a separate issue describing the limitations you see in stamping as it's currently implemented, so that people more knowledgeable about that part of Bazel can help brainstorm potential solutions. If we determine that scrubbing is the appropriate way to solve it, I'm happy to resume the search for a reasonable implementation; but I'd first like to be sure that there aren't simpler solutions.

**Bazel 7 scrubbing configuration limitations**

I completely agree with your assessment: scrubbing configurations are both hard to write and hard to debug.

The second problem is somewhat easier to solve: we need good tools to investigate the root cause of cache misses. We're currently trying to make some improvements in that direction in https://github.com/bazelbuild/bazel/issues/18643, but there's still work to do.

The first problem, as you correctly surmise, doesn't have an easy solution in sight. I share your sentiment that the better design entails having the rules themselves provide the scrubbing configuration to Bazel, so that users aren't forced to understand implementation details just so they can write the configuration. But we don't want rules to unilaterally decide that scrubbing should be used, either: scrubbing trades off correctness for performance, and Bazel should strive for correctness by default; some amount of user input would still be required in that world.

**In conclusion**

I don't want this discussion to drag on forever. I offer the following: if we agree that the first of these problems is worth solving on its own (in other words: useful to you), I volunteer to (re-)write the PR myself and get it submitted. I will tentatively say that there's still time to do this before 7.1.0, but it might slip to 7.2.0 since I have other things on my plate at the moment. Does that sound good to you?

Comment (bozaro) [reply src/main/java/com/google/devtools/build/lib/remote/RemoteExecutionService.java:345]: If you add the matcher `has_volatile_input` and allow remote assembly for all rules that do not use scrubbing, then this should completely cover the problem of using stamping from Bazel 6:
 - there will be one rule that allows local assembly of stamping action;
 - this rule will behave the same as the local Bazel cache.

Comment (bozaro) [reply src/main/java/com/google/devtools/build/lib/remote/RemoteExecutionService.java:345]: @tjgq, I have moved the changes about executing scrubbed actions locally to PR #21288 (check only by `scrubber.forSpawn(spawn) != null` without checking the affected input files).

I tried adding `has_volatile_input' but failed: this requires strong refactoring.

Comment (tjgq) [reply src/main/java/com/google/devtools/build/lib/remote/RemoteExecutionService.java:345]: Thanks, I will import https://github.com/bazelbuild/bazel/pull/21288 today.

The `has_volatile_input` implementation is still on me, as promised above; I'm hoping to get to it next week.

Comment (tjgq) [reply src/main/java/com/google/devtools/build/lib/remote/RemoteExecutionService.java:345]: I had some time this week to chat with other folks on the team about stamping, and to think about how this feature might be implemented. Unfortunately, the conclusion that I came to is that I don't know how to implement this in a satisfactory way, and I really don't want to rush adding a feature we might later regret. So I'm going to put this on hold at least until 7.2.0. Sorry.

The internal discussions I had also reinforced a point that I made earlier in this thread: if you have so many targets consuming `volatile-info.txt` that it's difficult to explicitly enumerate all of them in the scrubbing config, you're likely doing something wrong (at least according to the way we originally designed stamping to be used). I'd expect you to have at most one target per language that consumes `volatile-info.txt` and produces a wrapper library around its contents suitable for that language; your top-level binary targets should access the volatile information through that library, and shouldn't themselves depend on `volatile-info.txt` directly. Thus, you only have to configure scrubbing for the small number of targets that generate the per-language wrapper libraries, which I don't think is an unreasonable thing to ask users to do.

I still haven't read a convincing explanation for why your build graph can't be organized in this way.

**Comment by meisterT** [comment]
Closing as per https://github.com/bazelbuild/bazel/pull/20070#discussion_r1499534883
